QEP-label-path.rst

```diff
@@ -0,0 +1,71 @@
+.. _qep#[.#]:
+
+========================================================================
+QGIS Enhancement ??: Labeling path
+========================================================================
+
+:Date: October 2015
+:Author: Hugo Mercier
+:Contact: hugo dot mercier at oslandia dot com
+:Last Edited: 
+:Status:  
+:Version:
+
+Summary
+-------
+
+A new labeling option is proposed for addition here. It allows to use a user-defined linear path for the drawing of a label.
+
+Using a simplified path for the curved placement of labels is one of the possible use cases of such a feature.
+
+A user-defined path can be seen as an additional geometry on a layer, where the underlying model supports only one "geometry column".
+
+Proposed solution
+-----------------
+
+We propose the add a way for the user to use a different geometry object as support for the "curved" placement of a label.
+
+The current (2.11) version of QGIS allows to display labels along the linestring features in a "curved" manner.
+
+.. image:: curved_labeling.png
+
+However for linestrings defined with lots of control points (such as waterways) or with high curvatures, displaying correct
+labels can be either unesthetic or very slow.
+
+This is where using a path slightly different from the feature's geometry could be useful. For instance using a simplified version
+of the linestring could lead to a label faster to display or with smaller curvatures.
+
+We then propose to allow the "curved" placement to use an arbitrary geometry, through a data-defined "label path" property. The data-defined
+property could point to a user-defined column or to an expression.
+
+We also propose to add a "simplify()" function in core expression functions for that purpose.
+
+This option for placement along a user-defined path would also be useful for other types of features, like points and polygons.
+
+Core Changes
+------------
+
+A new data-defined labeling property of type QgsGeometry is added for the definition of the label path ("labeling/dataDefined/labelPath").
+QgsPalLayerSettings::registerFeature will be modified in consequence: if this new property is set and curved placement is enabled, it will be used as a support.
+
```

Comment:
Note that it's not sufficient to just use the alternative geometry when registering a label. Instead, labels needs to be able to register an optional "obstacle geometry" which can differ from the feature geometry. Otherwise the simplified/smoothed line feature would be used for the obstacle detection too, and could result on labels being placed over the rendered symbol for a line.

Note that I've got a nearly-complete implementation of this I hope to merge soon (allowing labels to have different obstacle geometry).